gtars/src/uniwig/mod.rs

```diff
@@ -1,16 +1,722 @@
+use std::io::{BufRead, BufReader, Read, Write};
+use std::path::Path;
+use std::fs::{File, OpenOptions};
+use std::error::Error;
+use clap::builder::OsStr;
+use flate2::read::GzDecoder;
+use ndarray::Array;
+use ndarray_npy::write_npy;
+
+
+pub mod cli;
+
+pub mod consts {
+    pub const UNIWIG_CMD: &str = "uniwig";
+
 }
+
+pub struct Chromosome {
+    chrom: String,
+    starts: Vec<i32>,
+    ends: Vec<i32>,
+}
+impl Clone for Chromosome {
+    fn clone(&self) -> Self {
+        Self {
+            chrom: self.chrom.clone(), // Clone the string
+            starts: self.starts.clone(), // Clone the vector
+            ends: self.ends.clone(),   // Clone the vector
+        }
+    }
+}
+
+
+// pub fn read_bed_map(combinedbedpath: &str){
+//
+//
+// }
+
+pub fn read_bed_vec(combinedbedpath: &str) -> Vec<Chromosome> {
+
+    let path = Path::new(combinedbedpath);
+
+    let file = File::open(path).unwrap();
+
+    let is_gzipped = path.extension().unwrap_or(&OsStr::from("bed")) == "gz";
+
+    // We must encapsulate in a box and use a dynamic Read trait so that either case could continue.
+    let reader: Box<dyn Read> = match is_gzipped {
+        true => Box::new(GzDecoder::new(file)),
+        false => Box::new(file),
+    };
+
+    let reader = BufReader::new(reader);
+
+    let mut chromosome = Chromosome{
+        chrom: "".to_string(),
+        starts: vec![],
+        ends: vec![],
+    };
+
+    let mut chromosome_vec: Vec<Chromosome> = Vec::new();
+
+    let mut chrom = String::new();
+
+    for line in reader.lines() {
+        //println!("Here is line{:?}", line);
+
+        // Must use a 2nd let statement to appease the borrow-checker
+        let line_string = line.unwrap();
+        let s = line_string.as_str();
+
+        let (parsed_chr, parsed_start, parsed_end) = parse_bed_file(s).unwrap();
+
+        if chrom.is_empty(){
+            // Initial chromosome
+            chromosome.chrom = String::from(parsed_chr.trim());
+            chrom = String::from(parsed_chr.trim());
+            chromosome.starts.push(parsed_start);
+            chromosome.ends.push(parsed_end);
+            continue;
+        }
+
+
+        if String::from(parsed_chr.trim()) != chrom{
+
+            // If the parsed chrom is not the same as the current, sort, and then push to vector
+            // then reset chromosome struct using the newest parsed_chr
+            chromosome.starts.sort_unstable();
+            chromosome.ends.sort_unstable();
+
+            chromosome_vec.push(chromosome.clone());
+
+            chromosome.chrom =String::from(parsed_chr.trim());
+            chrom = String::from(parsed_chr.trim());
+
+            chromosome.starts = vec![];
+            chromosome.ends = vec![]
+        }
+
+        chromosome.starts.push(parsed_start);
+        chromosome.ends.push(parsed_end);
+
+    }
+
+    // Is this final sort and push actually necessary?
+    chromosome.starts.sort_unstable();
+    chromosome.ends.sort_unstable();
```

Comment:
Also, why unstable? 😵‍💫

Reply:
From docs:

When applicable, unstable sorting is preferred because it is generally faster than stable sorting and it doesn’t allocate auxiliary memory. See [sort_unstable](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable).

https://doc.rust-lang.org/std/primitive.slice.html#method.sort